sesman/chansrv/clipboard.c

```diff
@@ -1148,20 +1158,35 @@ clipboard_process_data_response_for_image(struct stream *s,
     {
         return 0;
     }
+
     g_free(g_clip_c2s.data);
-    g_clip_c2s.data = (char *) g_malloc(len + 14, 0);
+    g_clip_c2s.data = (char *) g_malloc(len + BMPFILEHEADER_LEN, 0);
     if (g_clip_c2s.data == 0)
     {
         g_clip_c2s.total_bytes = 0;
         return 0;
     }
-    g_clip_c2s.total_bytes = len;
+    g_clip_c2s.total_bytes = len + BMPFILEHEADER_LEN;
     g_clip_c2s.read_bytes_done = g_clip_c2s.total_bytes;
-    g_memcpy(g_clip_c2s.data, g_bmp_image_header, 14);
-    in_uint8a(s, g_clip_c2s.data + 14, len);
+
+    /* assemble bmp header */
+    g_memcpy(bmp_file_header, g_bmp_image_header, BMPFILEHEADER_LEN);
+    bmp_size = (uint32_t) g_clip_c2s.total_bytes;
+#if defined(L_ENDIAN)
+    for (int i = 2, j = 0; i < 6; i++, j += 8)
+#else
+    for (int i = 5, j = 0; i >= 2; i--, j += 8)
+#endif
+    {
+        bmp_file_header[i] = ((bmp_size >> j) & 0xff);
```

Comment:
@matt335672 Can you have a look at this area especially? I'm not sure if this is a good way to embed file size in BMP file header.

Reply:
Yes, that could do with tidying up.

Personally, I'd use the stream functions we already have. They've been well-tested, an xrdp developer should be familiar with them, and a lot of magic numbers can be avoided. Something like this maybe (compiles, but not tested):-

```c
    /* Assemble bmp header */
    /* See https://en.wikipedia.org/wiki/BMP_file_format#Bitmap_file_header */
    make_stream(hdr_s);
    if (hdr_s == 0)
    {
        g_free(g_clip_c2s.data);
        g_clip_c2s.total_bytes = 0;
        return 0;
    }
    init_stream(hdr_s, BMPFILEHEADER_LEN);
    out_uint8(hdr_s, 'B');
    out_uint8(hdr_s, 'M');
    out_uint32_le(hdr_s, g_clip_c2s.total_bytes);
    out_uint16_le(hdr_s, 0);
    out_uint16_le(hdr_s, 0);
    out_uint32_le(hdr_s, BMPFILEHEADER_LEN + 40);

    /* Copy header and data to output stream */
    g_memcpy(g_clip_c2s.data, hdr_s->data, BMPFILEHEADER_LEN);
    in_uint8a(s, g_clip_c2s.data + BMPFILEHEADER_LEN, len);
    free_stream(hdr_s);
```
You can remove `g_bmp_image_header`, and simply define `BMPFILEHEADER_LEN` as 14.

It's not as efficient, but that probably doesn't matter. It's not being called that often to need to optimise it.

Reply:
Thanks, it worked and applied tidy-up. I also replaced the magic number `40` with `BMPINFOHEADER_LEN`.